### PR TITLE
Terra bullet (gdx native)

### PR DESCRIPTION
**Pull request by pollend**
This implementation uses a version of libgdx bullet extension to work directly with Termath along with a voxel implementation. there are some out standing features that need to be address and fixed before  this would be considered workable but this implementation works against the current development branch of Terasology.

https://github.com/MovingBlocks/Terasology/pull/3121
https://github.com/MovingBlocks/Terasology/pull/3042

https://github.com/pollend/TeraBullet-2

# How to Test

~~You will have to include the modified libgdx library into lib and install the maven libraries from there (mvn install). At the moment the way to load the native library is to just copy the native-build into the native folder in the root of the project. running the ant script in TeraMath should do everything for you in terms of building the library.~~

The bullet builds are in artifactory so there is no need to copy anything into lib.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

File: engine/src/main/java/org/terasology/logic/inventory/ItemPickupAuthoritySystem.java [hunk 1]
```diff
@@ -99,10 +99,8 @@ public void updateExtentsOnBlockItemBoxShape(OnAddedComponent event, EntityRef i
             return;
         }
 
-        if (blockFamily.getArchetypeBlock().getCollisionShape() instanceof BoxShape) {
-            BoxShape collisionShape = (BoxShape) blockFamily.getArchetypeBlock().getCollisionShape();
-            Vector3f extents = collisionShape.getHalfExtentsWithoutMargin();
-            extents.scale(2.0f);
+        if (blockFamily.getArchetypeBlock().getCollisionShape() instanceof btBoxShape) {
+            Vector3f extents = ((btBoxShape) blockFamily.getArchetypeBlock().getCollisionShape()).getHalfExtentsWithoutMargin();
```
Comment (eviltak) [inline engine/src/main/java/org/terasology/logic/inventory/ItemPickupAuthoritySystem.java]: After #3199, this will not work. Please revert this change.

File: engine/src/main/java/org/terasology/physics/bullet/shapes/BulletBoxShape.java [hunk 2]
```diff
@@ -15,36 +15,28 @@
...
-        return new BulletBoxShape(halfExtentsWithMargin);
+        Vector3f halfExtentsWithMargin = boxShape.getHalfExtentsWithMargin();
+        return new BulletBoxShape(rot.rotate(halfExtentsWithMargin));
     }
 
     @Override
     public Vector3f getHalfExtentsWithoutMargin() {
-        javax.vecmath.Vector3f out = new javax.vecmath.Vector3f();
-        return VecMath.from(boxShape.getHalfExtentsWithoutMargin(out));
+        return boxShape.getHalfExtentsWithoutMargin().mul(2);
```
Comment (eviltak) [inline engine/src/main/java/org/terasology/physics/bullet/shapes/BulletBoxShape.java]: I recommend that you leave out the `.mul(2)` for now since the method has not been renamed to `getExtents` (though you have already merged the commit that renames the method to `getExtents` from #3290 -- not sure what has happened here)

File: engine/src/main/java/org/terasology/physics/bullet/shapes/BulletCapsuleShape.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.physics.bullet.shapes;
+
+import com.badlogic.gdx.physics.bullet.collision.btCapsuleShape;
+import org.terasology.math.geom.Quat4f;
+import org.terasology.physics.shapes.CollisionShape;
+
+public class BulletCapsuleShape extends BulletCollisionShape {
```
Comment (eviltak) [inline engine/src/main/java/org/terasology/physics/bullet/shapes/BulletCapsuleShape.java]: Ideally this should also implement an engine `CapsuleShape` interface, but it looks like I may have forgotten to add the interface to the CollisionShape API. I will add it in #3290, after which it can be implemented here.

File: engine/src/main/java/org/terasology/physics/bullet/shapes/BulletCylinderShape.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+package org.terasology.physics.bullet.shapes;
+
+import com.badlogic.gdx.physics.bullet.collision.btConvexHullShape;
+import com.badlogic.gdx.physics.bullet.collision.btCylinderShape;
+import org.terasology.math.geom.Quat4f;
+import org.terasology.math.geom.Vector3f;
+import org.terasology.physics.shapes.CollisionShape;
+
+
+public class BulletCylinderShape extends  BulletCollisionShape{
```
Comment (eviltak) [inline engine/src/main/java/org/terasology/physics/bullet/shapes/BulletCylinderShape.java]: Like `CapsuleShape`, looks like the `CylinderShape` interface is missing too. I shall add it in #3290.

File: engine/src/main/java/org/terasology/physics/shapes/CollisionShapeFactory.java [hunk 5]
```diff
@@ -62,4 +64,13 @@ default BoxShape getNewUnitCube() {
...
     SphereShape getNewSphere(float radius);
+
+
+    /**
+     * Returns the colliding body associated with the entity
+     *
+     * @param entityRef
+     * @return
+     */
+    BulletCollisionShape getShapeFor(EntityRef entityRef);
```
Comment (eviltak) [inline engine/src/main/java/org/terasology/physics/shapes/CollisionShapeFactory.java]: Since a `Factory` is responsible for creating objects, I would prefer moving this method somewhere else. Besides, this method does not need to be overridable -- similar to your implementation in `BulletCollisionShapeFactory`, the final implementation will look for components on the entity and return shapes created from a `CollisionShapeFactory` instance. Also, in the interest of remaining physics engine agnostic, the return type should be made `CollisionShape`. Sample final implementation:

```java
CollisionShape getShapeFor(EntityRef entity, CollisionShapeFactory collisionShapeFactory) {
        BoxShapeComponent box = entityRef.getComponent(BoxShapeComponent.class);
        if (box != null) {
            Vector3f halfExtents = new Vector3f(box.extents);
            return collisionShapeFactory.getNewBox(halfExtents.mul(.5f));
        }
        SphereShapeComponent sphere = entityRef.getComponent(SphereShapeComponent.class);
        if (sphere != null) {
            return collisionShapeFactory.getNewSphere(sphere.radius);
        }
        CapsuleShapeComponent capsule = entityRef.getComponent(CapsuleShapeComponent.class);
        if (capsule != null) {
            // This method does not yet exist in CollisionShapeFactory
            return collisionShapeFactory.getNewCapsule(capsule.radius,capsule.height);
        }
        CylinderShapeComponent cylinder = entityRef.getComponent(CylinderShapeComponent.class);
        if (cylinder != null) {
            // This method does not yet exist in CollisionShapeFactory
            return collisionShapeFactory.getNewCylinder(new Vector3f(cylinder.radius, 0.5f * cylinder.height, cylinder.radius));
        }
        HullShapeComponent hull = entityRef.getComponent(HullShapeComponent.class);
        if (hull != null) {
            FloatBuffer buffer = BufferUtils.createFloatBuffer( hull.sourceMesh.getVertices().size());
            TFloatIterator iterator = hull.sourceMesh.getVertices().iterator();
            int numPoints = 0;
            while (iterator.hasNext()) {
                numPoints++;
                buffer.put(iterator.next());
                buffer.put(iterator.next());
                buffer.put(iterator.next());
            }
            return collisionShapeFactory.getNewConvexHull(buffer,numPoints,3 * Float.BYTES);
        }
        CharacterMovementComponent characterMovementComponent = entityRef.getComponent(CharacterMovementComponent.class);
        if (characterMovementComponent != null) {
            return collisionShapeFactory.getNewCapsule(characterMovementComponent.radius,characterMovementComponent.height);
        }
        throw new IllegalArgumentException("Creating physics object that requires a ShapeComponent or CharacterMovementComponent, but has neither. Entity: " + entityRef);
    }
```

File: engine/src/main/java/org/terasology/physics/bullet/shapes/BulletCollisionShapeFactory.java [hunk 6]
```diff
@@ -44,4 +58,44 @@ public CompoundShape getNewCompoundShape() {
     public SphereShape getNewSphere(float radius) {
         return new BulletSphereShape(radius);
     }
+
+    @Override
+    public BulletCollisionShape getShapeFor(EntityRef entityRef) {
```
Comment (eviltak) [inline engine/src/main/java/org/terasology/physics/bullet/shapes/BulletCollisionShapeFactory.java]: Like I explain below, I think this method should be moved elsewhere.

File: engine/src/main/java/org/terasology/physics/bullet/shapes/BulletCollisionShapeFactory.java [hunk 7]
```diff
@@ -44,4 +58,44 @@ public CompoundShape getNewCompoundShape() {
...
+                buffer.put(iterator.next());
+            }
+            return new BulletConvexHullShape(buffer,numPoints,3 * Float.BYTES);
+        }
+        CharacterMovementComponent characterMovementComponent = entityRef.getComponent(CharacterMovementComponent.class);
+        if (characterMovementComponent != null) {
+            return new BulletCapsuleShape(characterMovementComponent.radius,characterMovementComponent.height);
+        }
+        logger.error("Creating physics object that requires a ShapeComponent or CharacterMovementComponent, but has neither. Entity: {}", entityRef);
+        throw new IllegalArgumentException("Creating physics object that requires a ShapeComponent or CharacterMovementComponent, but has neither. Entity: " + entityRef);
```
Comment (eviltak) [inline engine/src/main/java/org/terasology/physics/bullet/shapes/BulletCollisionShapeFactory.java]: If we are already throwing an exception, a logger error may be redundant.

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/BaseLwjglSubsystem.java [hunk 8]
```diff
@@ -38,6 +39,7 @@ public void preInitialise(Context context) {
         if (!initialised) {
             initLogger();
             LWJGLHelper.initNativeLibs();
+            Bullet.init(true);
```
Comment (eviltak) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/BaseLwjglSubsystem.java]: The introduction of this bullet-specific line of code in the engine directly breaks the independence of the engine from the physics engine. If calling this method is absolutely necessary, we may have to look at an abstraction for physics engine initialization (like we did with the CollisionShape API). This could probably be done in `PhysicsEngineManager`, as that is what currently sets up bullet as the game physics engine.

File: engine/src/main/java/org/terasology/physics/engine/RigidBody.java [hunk 9]
```diff
@@ -120,6 +121,9 @@
      */
     void setOrientation(Quat4f orientation);
 
+    Matrix4f getWorldTransform();
+    Matrix4f setWorldTransform(Matrix4f trans);
```
Comment (eviltak) [inline engine/src/main/java/org/terasology/physics/engine/RigidBody.java]: Will it be better to use the `Transform` class here (or modify `Transform` to be created from/use `Matrix4f`) instead?

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by pollend** [comment]
the steps in the original PR are obsolete. the artifacts are now in artifactory so the basic steps just involve running the project.

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that
